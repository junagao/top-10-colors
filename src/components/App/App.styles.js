import styled from '@emotion/styled'
import theme from 'utils/theme'

export const Container = styled.div`
  max-width: 80vw;
  margin: 0 auto;
`

export const Header = styled.h1`
  color: ${theme.colors.primary};
  font-size: ${theme.fontSizes.lg};
  padding: ${theme.space[5]} ${theme.space[0]};
`
