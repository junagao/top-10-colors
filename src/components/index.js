export { default as ColorList } from 'components/ColorList'
export { default as ColorItem } from 'components/ColorItem'
export { default as Header } from 'components/Header'
export { default as Rating } from 'components/Rating'
export { default as Star } from 'components/Star'
