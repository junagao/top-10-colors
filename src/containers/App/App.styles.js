import styled from '@emotion/styled'

const AppContainer = styled.div`
  max-width: 80vw;
  margin: 0 auto;
  @media (min-width: 1920px) {
    max-width: 60vw;
  }
`

export default AppContainer
