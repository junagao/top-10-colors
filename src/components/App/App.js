import React from 'react'
import { hot } from 'react-hot-loader/root'
import { ColorList } from 'components'
import { Container, Header } from './App.styles'

const App = () => (
  <Container>
    <Header>Top 10 favorite colors</Header>
    <ColorList />
  </Container>
)

export default hot(App)
