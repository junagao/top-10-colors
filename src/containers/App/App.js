import React from 'react'
import { arrayOf, shape, func, bool } from 'prop-types'
import { hot } from 'react-hot-loader/root'
import { connect } from 'react-redux'
import {
  getColors,
  rateColor,
  randomRateColor,
  toggleRandomRating,
  resetRatings,
} from 'actions/colors'
import { Header, ColorList } from 'components'
import AppContainer from './App.styles'

let ratingTimer = null
class App extends React.Component {
  componentDidMount() {
    const { getColors } = this.props
    getColors()
  }

  handleRating = (id, rating) => {
    const { colors, rateColor } = this.props
    const color = colors.find((color) => color.id === id)
    return color.rating === rating ? rateColor(id, 0) : rateColor(id, rating)
  }

  handleResetRatings = () => {
    const { resetRatings } = this.props
    resetRatings()
  }

  startRandomRating = () => {
    const { randomRateColor, toggleRandomRating } = this.props
    toggleRandomRating()

    const minDelay = 500
    const maxDelay = 1000

    const randomIndex = (min, max) =>
      Math.floor(Math.random() * (max - min + 1)) + min

    const randomDelay = () => randomIndex(minDelay, maxDelay)

    const randomRating = () => randomIndex(1, 5)
    const randomColorIndex = () => randomIndex(0, 9)

    ratingTimer = setInterval(
      () => randomRateColor(randomColorIndex(), randomRating()),
      randomDelay(),
    )
  }

  stopRandomRating = () => {
    const { toggleRandomRating } = this.props
    clearInterval(ratingTimer)
    toggleRandomRating()
  }

  render() {
    const { colors, isRandomRating } = this.props

    const sortedColors = colors.sort((a, b) => b.rating - a.rating)

    return (
      <AppContainer>
        <Header
          isRandomRating={isRandomRating}
          startRandomRating={this.startRandomRating}
          stopRandomRating={this.stopRandomRating}
          onResetRatings={this.handleResetRatings}
        />
        <ColorList
          colors={sortedColors}
          onRate={this.handleRating}
          onHover={this.handleHovering}
        />
      </AppContainer>
    )
  }
}

App.propTypes = {
  colors: arrayOf(shape({})).isRequired,
  getColors: func.isRequired,
  rateColor: func.isRequired,
  isRandomRating: bool.isRequired,
  randomRateColor: func.isRequired,
  toggleRandomRating: func.isRequired,
  resetRatings: func.isRequired,
}

const mapStateToProps = (state) => ({
  colors: state.colors,
  isRandomRating: state.isRandomRating,
})

const mapDispatchToProps = {
  getColors,
  rateColor,
  randomRateColor,
  toggleRandomRating,
  resetRatings,
}

export default hot(connect(mapStateToProps, mapDispatchToProps)(App))
