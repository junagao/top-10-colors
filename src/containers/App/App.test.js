import React from 'react'
import { shallow } from 'enzyme'
import { Header, ColorList } from 'components'
import { App } from './App'

describe('App', () => {
  let wrapper

  const mockColors = [
    {
      id: 'fc352593-c91d-4cf6-b689-7478ee2b61c8',
      name: 'orange',
      hex: '#FFA500',
      rgb: 'rgb(255, 165, 0)',
      rating: 3,
    },
    {
      id: 'b0e3fab2-0477-4e28-b2e0-c2281a21d09b',
      name: 'purple',
      hex: '#800080',
      rgb: 'RGB(128, 0, 128)',
      rating: 5,
    },
  ]

  beforeEach(() => {
    wrapper = shallow(
      <App
        colors={mockColors}
        getColors={jest.fn()}
        rateColor={jest.fn()}
        isRandomRating
        toggleRandomRating={jest.fn()}
        randomRateColor={jest.fn()}
        resetRatings={jest.fn()}
      />,
    )
  })

  it('should render correctly given the required props', () => {
    expect(wrapper).toMatchSnapshot()
  })

  it('renders one AppContainer element', () => {
    console.log(wrapper.debug())
    expect(wrapper.find('AppContainer').length).toEqual(1)
  })

  it('shows one instance of Header', () => {
    expect(wrapper.find(Header).length).toEqual(1)
  })

  it('shows one instance of ColorList', () => {
    expect(wrapper.find(ColorList).length).toEqual(1)
  })
})
