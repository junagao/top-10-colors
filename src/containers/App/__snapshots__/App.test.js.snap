// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should render correctly given the required props 1`] = `
<AppContainer>
  <Header
    isRandomRating={true}
    onResetRatings={[Function]}
    startRandomRating={[Function]}
    stopRandomRating={[Function]}
  />
  <ColorList
    colors={
      Array [
        Object {
          "hex": "#800080",
          "id": "b0e3fab2-0477-4e28-b2e0-c2281a21d09b",
          "name": "purple",
          "rating": 5,
          "rgb": "RGB(128, 0, 128)",
        },
        Object {
          "hex": "#FFA500",
          "id": "fc352593-c91d-4cf6-b689-7478ee2b61c8",
          "name": "orange",
          "rating": 3,
          "rgb": "rgb(255, 165, 0)",
        },
      ]
    }
    onRate={[Function]}
  />
</AppContainer>
`;
